File: enode/src/main/java/org/enodeframework/domain/impl/DefaultAggregateRepositoryProvider.java
package org.enodeframework.domain.impl;

import org.enodeframework.common.container.ObjectContainer;
import org.enodeframework.domain.IAggregateRepository;
import org.enodeframework.domain.IAggregateRepositoryProvider;
import org.enodeframework.domain.IAggregateRepositoryProxy;
import org.enodeframework.domain.IAggregateRoot;
import org.enodeframework.infrastructure.IAssemblyInitializer;
import org.enodeframework.infrastructure.TypeUtils;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * @author anruence@gmail.com
 */
public class DefaultAggregateRepositoryProvider implements IAggregateRepositoryProvider, IAssemblyInitializer {

    private final Map<Class<?>, IAggregateRepositoryProxy> repositoryDict = new HashMap<>();

    @Override
    public IAggregateRepositoryProxy getRepository(Class<? extends IAggregateRoot> aggregateRootType) {
        return repositoryDict.get(aggregateRootType);
    }

    @Override
    public void initialize(Set<Class<?>> componentTypes) {
        componentTypes.stream().filter(TypeUtils::isAggregateRepositoryType).forEach(this::registerAggregateRepository);
    }

    /**
     * 获取继承AggregateRoot的class，IAggregateRepository接口的泛型
     */
    private void registerAggregateRepository(Class<?> aggregateRepositoryType) {
        Type[] genericInterfaces = aggregateRepositoryType.getGenericInterfaces();
        Arrays.stream(genericInterfaces).forEach(x -> {
            ParameterizedType superGenericInterfaceType = (ParameterizedType) x;
            if (!IAggregateRepository.class.equals(superGenericInterfaceType.getRawType())) {
                return;
            }
            AggregateRepositoryProxy<IAggregateRoot> aggregateRepositoryProxy = new AggregateRepositoryProxy<>();
            aggregateRepositoryProxy.setInnerObject(ObjectContainer.resolve(aggregateRepositoryType));
            repositoryDict.put((Class<?>) superGenericInterfaceType.getActualTypeArguments()[0], aggregateRepositoryProxy);
        });
    }
}

File: enode/src/main/java/org/enodeframework/domain/impl/DefaultAggregateRepositoryProvider.kt
package org.enodeframework.domain.impl

import org.enodeframework.common.container.ObjectContainer
import org.enodeframework.common.exception.AggregateRootTypeRepositoryNotFoundException
import org.enodeframework.domain.AggregateRepository
import org.enodeframework.domain.AggregateRepositoryProvider
import org.enodeframework.domain.AggregateRepositoryProxy
import org.enodeframework.domain.AggregateRoot
import org.enodeframework.infrastructure.AssemblyInitializer
import org.enodeframework.infrastructure.TypeUtils
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * @author anruence@gmail.com
 */
class DefaultAggregateRepositoryProvider : AggregateRepositoryProvider, AssemblyInitializer {
    private val repositoryDict: MutableMap<Class<*>, AggregateRepositoryProxy> = HashMap()
    override fun getRepository(aggregateRootType: Class<out AggregateRoot>): AggregateRepositoryProxy {
        return repositoryDict[aggregateRootType]
            ?: throw AggregateRootTypeRepositoryNotFoundException(aggregateRootType.name)
    }

    override fun initialize(objectContainer: ObjectContainer, componentTypes: Set<Class<*>>) {
        componentTypes.filter { type: Class<*> -> TypeUtils.isAggregateRepositoryType(type) }
            .forEach { aggregateRepositoryType: Class<*> ->
                registerAggregateRepository(
                    objectContainer,
                    aggregateRepositoryType
                )
            }
    }

    /**
     * 获取继承AggregateRoot的class，IAggregateRepository接口的泛型
     */
    private fun registerAggregateRepository(objectContainer: ObjectContainer, aggregateRepositoryType: Class<*>) {
        aggregateRepositoryType.genericInterfaces.forEach { type: Type ->
            val superGenericInterfaceType = type as ParameterizedType
            if (AggregateRepository::class.java != superGenericInterfaceType.rawType) {
                return
            }
            val aggregateRepositoryProxy = DefaultAggregateRepositoryProxy<AggregateRoot>()
            aggregateRepositoryProxy.setInnerObject(objectContainer.resolve(aggregateRepositoryType))
            repositoryDict[superGenericInterfaceType.actualTypeArguments[0] as Class<*>] = aggregateRepositoryProxy
        }
    }
}